File: scripts/seed.ts
import { spawnSync } from 'child_process'
import { v4 as uuidv4 } from 'uuid'
import bcrypt from 'bcryptjs'

const runSql = (sql: string) => {
  console.log(`Executing: ${sql.substring(0, 100)}...`)
  const result = spawnSync('team-db', [sql], { encoding: 'utf-8' })
  if (result.error) {
    console.error('Error executing SQL:', result.error)
    return null
  }
  if (result.status !== 0) {
    console.error('SQL Error:', result.stderr)
    return null
  }
  try {
    return JSON.parse(result.stdout)
  } catch (e) {
    return result.stdout
  }
}

export const seed = async () => {
  console.log('Starting seeding...')

  const aliceId = uuidv4()
  const bobId = uuidv4()
  const charlieId = uuidv4()

  // Seed Users
  const hashedPass = await bcrypt.hash('password123', 10)
  console.log('Hashed Pass:', hashedPass)
  runSql(`INSERT INTO users (id, email, name, password_hash, role, credit_score) VALUES ('${aliceId}', 'alice@example.com', 'Alice Seller', '${hashedPass}', 'seller', 750);`)
  runSql(`INSERT INTO users (id, email, name, password_hash, role, credit_score) VALUES ('${bobId}', 'bob@example.com', 'Bob Buyer', '${hashedPass}', 'buyer', 620);`)
  runSql(`INSERT INTO users (id, email, name, password_hash, role, credit_score) VALUES ('${charlieId}', 'charlie@example.com', 'Charlie Dealer', '${hashedPass}', 'seller', 800);`)

  const car1Id = uuidv4()
  const car2Id = uuidv4()

  // Seed Cars
  runSql(`INSERT INTO cars (id, seller_id, make, model, year, price, mileage, description, image_url, status, flexible_payment) VALUES ('${car1Id}', '${aliceId}', 'Toyota', 'Camry', 2020, 22000, 35000, 'Reliable Camry', 'https://example.com/camry.jpg', 'available', 1);`)
  runSql(`INSERT INTO cars (id, seller_id, make, model, year, price, mileage, description, image_url, status, flexible_payment) VALUES ('${car2Id}', '${charlieId}', 'Honda', 'Civic', 2018, 18000, 50000, 'Sporty Civic', 'https://example.com/civic.jpg', 'available', 1);`)

  console.log('Seeding completed.')
}

if (import.meta.url.endsWith(process.argv[1])) {
  seed()
}

File: db/seed.py
import subprocess
import json
import uuid
from datetime import datetime, timedelta

def run_sql(sql):
    print(f"Executing: {sql[:100]}...")
    result = subprocess.run(['team-db', sql], capture_output=True, text=True)
    if result.returncode != 0:
        print(f"Error: {result.stderr}")
        return None
    return json.loads(result.stdout)

def seed():
    # 1. Users
    users = [
        {
            "id": str(uuid.uuid4()),
            "name": "Alice Seller",
            "email": "alice@example.com",
            "password_hash": "hashed_password_1",
            "role": "seller",
            "credit_score": 750
        },
        {
            "id": str(uuid.uuid4()),
            "name": "Bob Buyer",
            "email": "bob@example.com",
            "password_hash": "hashed_password_2",
            "role": "buyer",
            "credit_score": 620
        },
        {
            "id": str(uuid.uuid4()),
            "name": "Charlie Dealer",
            "email": "charlie@example.com",
            "password_hash": "hashed_password_3",
            "role": "seller",
            "credit_score": 800
        }
    ]

    for user in users:
        run_sql(f"INSERT INTO users (id, name, email, password_hash, role, credit_score) VALUES ('{user['id']}', '{user['name']}', '{user['email']}', '{user['password_hash']}', '{user['role']}', {user['credit_score']});")

    # 2. Listings
    listings = [
        {
            "id": str(uuid.uuid4()),
            "seller_id": users[0]["id"],
            "make": "Toyota",
            "model": "Camry",
            "year": 2020,
            "mileage": 35000,
            "price": 22000,
            "description": "Reliable family car in great condition.",
            "images_urls": json.dumps(["https://example.com/camry1.jpg"]),
            "status": "active"
        },
        {
            "id": str(uuid.uuid4()),
            "seller_id": users[2]["id"],
            "make": "Honda",
            "model": "Civic",
            "year": 2018,
            "mileage": 50000,
            "price": 18000,
            "description": "Sporty and fuel-efficient.",
            "images_urls": json.dumps(["https://example.com/civic1.jpg"]),
            "status": "active"
        },
        {
            "id": str(uuid.uuid4()),
            "seller_id": users[0]["id"],
            "make": "Ford",
            "model": "F-150",
            "year": 2015,
            "mileage": 85000,
            "price": 25000,
            "description": "Tough truck for work or play.",
            "images_urls": json.dumps(["https://example.com/f1501.jpg"]),
            "status": "pending"
        }
    ]

    for listing in listings:
        run_sql(f"INSERT INTO listings (id, seller_id, make, model, year, mileage, price, description, images_urls, status) VALUES ('{listing['id']}', '{listing['seller_id']}', '{listing['make']}', '{listing['model']}', {listing['year']}, {listing['mileage']}, {listing['price']}, '{listing['description']}', '{listing['images_urls']}', '{listing['status']}');")

    # 3. Transactions
    transactions = [
        {
            "id": str(uuid.uuid4()),
            "listing_id": listings[2]["id"],
            "buyer_id": users[1]["id"],
            "seller_id": users[0]["id"],
            "amount": 25000,
            "type": "installment",
            "status": "completed"
        }
    ]

    for tx in transactions:
        run_sql(f"INSERT INTO transactions (id, listing_id, buyer_id, seller_id, amount, type, status) VALUES ('{tx['id']}', '{tx['listing_id']}', '{tx['buyer_id']}', '{tx['seller_id']}', {tx['amount']}, '{tx['type']}', '{tx['status']}');")

    # 4. Installment Plans
    plans = [
        {
            "id": str(uuid.uuid4()),
            "transaction_id": transactions[0]["id"],
            "total_amount": 25000,
            "down_payment": 5000,
            "monthly_payment": 600,
            "term_months": 36,
            "status": "active",
            "next_due_date": (datetime.now() + timedelta(days=30)).strftime('%Y-%m-%d %H:%M:%S')
        }
    ]

    for plan in plans:
        run_sql(f"INSERT INTO installment_plans (id, transaction_id, total_amount, down_payment, monthly_payment, term_months, status, next_due_date) VALUES ('{plan['id']}', '{plan['transaction_id']}', {plan['total_amount']}, {plan['down_payment']}, {plan['monthly_payment']}, {plan['term_months']}, '{plan['status']}', '{plan['next_due_date']}');")

    # 5. Payments
    payments = [
        {
            "id": str(uuid.uuid4()),
            "installment_plan_id": plans[0]["id"],
            "amount": 5000,
            "due_date": datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
            "paid_date": datetime.now().strftime('%Y-%m-%d %H:%M:%S'),
            "status": "completed"
        }
    ]

    for p in payments:
        run_sql(f"INSERT INTO payments (id, installment_plan_id, amount, due_date, paid_date, status) VALUES ('{p['id']}', '{p['installment_plan_id']}', {p['amount']}, '{p['due_date']}', '{p['paid_date']}', '{p['status']}');")

if __name__ == "__main__":
    seed()
